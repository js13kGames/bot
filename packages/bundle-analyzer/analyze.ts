import { unzip } from "./unzip";
import { parse as parseUrl } from "url";
import * as path from "path";
import { getHash } from "./md5";
import { upload } from "./s3";
import { runGame } from "./browserstack";
import { promisify } from "util";
import { Rules } from "js13kGames-bot-rules";
import getPixels from "get-pixels";
import { checkDescriptions, Result, CheckId } from "./checks";

export const analyze = async (rules: Rules, bundleContent: Buffer) => {
  const key = getHash(bundleContent);
  const report = createInitialReport();

  report.checks.bundle_size.result =
    bundleContent.length <= rules.bundle.max_size ? "ok" : "failed";

  // unzip
  let files: ReturnType<typeof unzip>;
  try {
    files = unzip(bundleContent);
    report.checks.bundle_valid_zip.result = "ok";
  } catch (error: any) {
    report.checks.bundle_valid_zip.result = "failed";
    report.checks.bundle_valid_zip.details = error.message;
    return report;
  }

  // check index.html
  {
    const ok = files.some((x) => x.filename === "index.html");

    if (ok) {
      report.checks.bundle_contains_index.result = "ok";
    } else {
      report.checks.bundle_contains_index.result = "failed";

      const htmlFile = files.find((x) => x.filename.endsWith(".html"));

      if (htmlFile && htmlFile.filename.endsWith("index.html"))
        report.checks.bundle_contains_index.details = `Can you try moving "${htmlFile.filename}" to the root of the zip ?`;
      if (htmlFile && !htmlFile.filename.endsWith("index.html"))
        report.checks.bundle_contains_index.details = `Can you try renaming "${htmlFile.filename}" to "index.html" ?`;
    }
  }

  // deploy
  let gameUrl = await uploadFiles(key, files);

  if (!gameUrl) {
    return report;
  } else {
    report.deployUrl = gameUrl;
  }

  // test game
  const { urls, errorlogs, base64screenShot } = await runGame(gameUrl);

  /// check for errors
  {
    report.checks.game_no_error.result = errorlogs.length ? "failed" : "ok";

    if (errorlogs.length)
      report.checks.game_no_error.details =
        `got ${errorlogs.length} errors:\n` +
        errorlogs.map((message) => `  ${message}`).join("\n");
  }

  // check for forbidden requests
  {
    const [baseUrl] = gameUrl.split("/index.html");
    const externalUrls = urls
      .filter((url) => !url.startsWith(baseUrl))
      .filter(
        (url) =>
          !rules.game.http_request_whitelist_regexp.some((re) =>
            url.match(new RegExp(re))
          )
      );

    report.checks.game_no_external_http.result = externalUrls.length
      ? "failed"
      : "ok";

    if (externalUrls.length)
      report.checks.game_no_external_http.details =
        `got ${externalUrls.length} forbidden requests:\n` +
        externalUrls
          .map((url) => {
            const hostname = parseUrl(url).hostname;
            const baseHostname = parseUrl(baseUrl).hostname;

            if (hostname && hostname === baseHostname)
              return `- ${
                url.split(hostname)[1]
              } Can you try importing this resource with a relative path ?`;

            return `- ${url}`;
          })
          .join("\n");
  }

  // check for blank screen
  {
    const getPixelsP = promisify(getPixels);
    const { data: dataImage } = await getPixelsP(
      "data:image/png;base64," + base64screenShot,
      ""
    );

    report.checks.game_no_blank_screen.result = isImageBlank(dataImage)
      ? "failed"
      : "ok";
  }

  return report;
};

const uploadFiles = async (key: string, files: any[]) => {
  let indexUrl: string | undefined;
  await Promise.all(
    files.map(({ filename, content }) =>
      upload(path.join(key, filename), content).then((url) => {
        if (filename === "index.html") indexUrl = url;
      })
    )
  );
  return indexUrl;
};

// return true if the image is filled with only one color
const isImageBlank = (data: ArrayLike<number>) => {
  let err = 0;

  for (let i = 0; i < data.length; i += 4) {
    err +=
      Math.abs(data[i + 0] - data[0]) +
      Math.abs(data[i + 1] - data[1]) +
      Math.abs(data[i + 2] - data[2]);
  }

  return err === 0;
};

const createInitialReport = () => {
  const checks: any = {};

  for (const id in checkDescriptions) {
    checks[id] = { result: "untested" };
  }

  return {
    deployUrl: null as null | string,
    checks: checks as Record<CheckId, { result: Result; details?: string }>,
  };
};

export const formatReport = (
  report: ReturnType<typeof createInitialReport>
) => ({
  ...report,
  checks: Object.entries(report.checks).map(([id, value]) => ({
    id,
    ...value,
    description: checkDescriptions[id as CheckId],
  })),
});
