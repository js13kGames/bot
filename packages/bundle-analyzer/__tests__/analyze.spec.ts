import * as fs from "fs";
import * as path from "path";
import { analyze } from "../analyze";

const FIXTURE_DIR = path.resolve(__dirname, "..", "__fixtures__");

jest.setTimeout(50000);

for (const file of fs.readdirSync(FIXTURE_DIR))
  it(`should analyze ${file}`, async () => {
    const rules = {
      bundle: { max_size: 13 * 1024 },
      game: { http_request_whitelist: [] },
    };

    const zip = fs.readFileSync(path.resolve(FIXTURE_DIR, file, "bundle.zip"));

    const report = await analyze(rules, zip);
    const genericReport = removeEnvSpecificData(report);

    genericReport.sha256SignatureOfTheMd5Hash =
      genericReport.sha256SignatureOfTheMd5Hash && "<signature>";

    expect(genericReport).toMatchSnapshot();
  });

// redact the actual bucket name from the report
// so the report is the same regardless of the bucket name
const removeEnvSpecificData = (o: Object) =>
  JSON.parse(
    JSON.stringify(o).replace(
      new RegExp(process.env.AWS_BUCKET_NAME!, "g"),
      "<bucketName>"
    )
  );
