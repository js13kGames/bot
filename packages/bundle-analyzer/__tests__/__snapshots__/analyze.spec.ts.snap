// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should analyze absolute-import-path 1`] = `
Object {
  "checks": Object {
    "bundle_contains_index": Object {
      "result": "ok",
    },
    "bundle_size": Object {
      "result": "ok",
    },
    "bundle_valid_zip": Object {
      "result": "ok",
    },
    "game_no_blank_screen": Object {
      "result": "ok",
    },
    "game_no_error": Object {
      "details": "got 2 errors:
  http://<bucketName>.s3-website-eu-west-1.amazonaws.com/asset/image/1659821.jpeg - Failed to load resource: the server responded with a status of 404 (Not Found)
  http://<bucketName>.s3-website-eu-west-1.amazonaws.com/asset/script.js - Failed to load resource: the server responded with a status of 404 (Not Found)",
      "result": "failed",
    },
    "game_no_external_http": Object {
      "details": "got 2 forbidden requests:
- /asset/image/1659821.jpeg Can you try importing this resource with a relative path ?
- /asset/script.js Can you try importing this resource with a relative path ?",
      "result": "failed",
    },
  },
  "deployUrl": "http://<bucketName>.s3-website-eu-west-1.amazonaws.com/b2ed77c919951cd7c6be1eebec67ba70/index.html",
}
`;

exports[`should analyze blank-screen 1`] = `
Object {
  "checks": Object {
    "bundle_contains_index": Object {
      "result": "ok",
    },
    "bundle_size": Object {
      "result": "ok",
    },
    "bundle_valid_zip": Object {
      "result": "ok",
    },
    "game_no_blank_screen": Object {
      "result": "failed",
    },
    "game_no_error": Object {
      "result": "ok",
    },
    "game_no_external_http": Object {
      "result": "ok",
    },
  },
  "deployUrl": "http://<bucketName>.s3-website-eu-west-1.amazonaws.com/6dd4e6691e2ead04a88ee4dc99a9ed99/index.html",
}
`;

exports[`should analyze error-on-run 1`] = `
Object {
  "checks": Object {
    "bundle_contains_index": Object {
      "result": "ok",
    },
    "bundle_size": Object {
      "result": "ok",
    },
    "bundle_valid_zip": Object {
      "result": "ok",
    },
    "game_no_blank_screen": Object {
      "result": "ok",
    },
    "game_no_error": Object {
      "details": "got 1 errors:
  http://<bucketName>.s3-website-eu-west-1.amazonaws.com/b3107c5e2f9240efdff74a17a6e5ef0b/asset/script.js 7:10 Uncaught Error: something bad",
      "result": "failed",
    },
    "game_no_external_http": Object {
      "result": "ok",
    },
  },
  "deployUrl": "http://<bucketName>.s3-website-eu-west-1.amazonaws.com/b3107c5e2f9240efdff74a17a6e5ef0b/index.html",
}
`;

exports[`should analyze external-http-call-html 1`] = `
Object {
  "checks": Object {
    "bundle_contains_index": Object {
      "result": "ok",
    },
    "bundle_size": Object {
      "result": "ok",
    },
    "bundle_valid_zip": Object {
      "result": "ok",
    },
    "game_no_blank_screen": Object {
      "result": "ok",
    },
    "game_no_error": Object {
      "result": "ok",
    },
    "game_no_external_http": Object {
      "details": "got 2 forbidden requests:
- https://avatars0.githubusercontent.com/u/1659820?v=4
- https://github.com/platane.png",
      "result": "failed",
    },
  },
  "deployUrl": "http://<bucketName>.s3-website-eu-west-1.amazonaws.com/c0d13b09ee4ae30c8938235cbbb7ae2e/index.html",
}
`;

exports[`should analyze index-not-in-root 1`] = `
Object {
  "checks": Object {
    "bundle_contains_index": Object {
      "details": "Can you try moving \\"game/index.html\\" to the root of the zip ?",
      "result": "failed",
    },
    "bundle_size": Object {
      "result": "ok",
    },
    "bundle_valid_zip": Object {
      "result": "ok",
    },
    "game_no_blank_screen": Object {
      "result": "untested",
    },
    "game_no_error": Object {
      "result": "untested",
    },
    "game_no_external_http": Object {
      "result": "untested",
    },
  },
  "deployUrl": null,
}
`;

exports[`should analyze missing-index 1`] = `
Object {
  "checks": Object {
    "bundle_contains_index": Object {
      "details": "Can you try renaming \\"game.html\\" to \\"index.html\\" ?",
      "result": "failed",
    },
    "bundle_size": Object {
      "result": "ok",
    },
    "bundle_valid_zip": Object {
      "result": "ok",
    },
    "game_no_blank_screen": Object {
      "result": "untested",
    },
    "game_no_error": Object {
      "result": "untested",
    },
    "game_no_external_http": Object {
      "result": "untested",
    },
  },
  "deployUrl": null,
}
`;

exports[`should analyze oversize-bundle 1`] = `
Object {
  "checks": Object {
    "bundle_contains_index": Object {
      "result": "ok",
    },
    "bundle_size": Object {
      "result": "failed",
    },
    "bundle_valid_zip": Object {
      "result": "ok",
    },
    "game_no_blank_screen": Object {
      "result": "ok",
    },
    "game_no_error": Object {
      "result": "ok",
    },
    "game_no_external_http": Object {
      "result": "ok",
    },
  },
  "deployUrl": "http://<bucketName>.s3-website-eu-west-1.amazonaws.com/8007d43b7ccd9e60484608a3d6be6b19/index.html",
}
`;

exports[`should analyze valid 1`] = `
Object {
  "checks": Object {
    "bundle_contains_index": Object {
      "result": "ok",
    },
    "bundle_size": Object {
      "result": "ok",
    },
    "bundle_valid_zip": Object {
      "result": "ok",
    },
    "game_no_blank_screen": Object {
      "result": "ok",
    },
    "game_no_error": Object {
      "result": "ok",
    },
    "game_no_external_http": Object {
      "result": "ok",
    },
  },
  "deployUrl": "http://<bucketName>.s3-website-eu-west-1.amazonaws.com/cec73ecd46375ab7c64a85c8a4d2c172/index.html",
}
`;

exports[`should analyze valid-whitelisted-external 1`] = `
Object {
  "checks": Object {
    "bundle_contains_index": Object {
      "result": "ok",
    },
    "bundle_size": Object {
      "result": "ok",
    },
    "bundle_valid_zip": Object {
      "result": "ok",
    },
    "game_no_blank_screen": Object {
      "result": "ok",
    },
    "game_no_error": Object {
      "result": "ok",
    },
    "game_no_external_http": Object {
      "result": "ok",
    },
  },
  "deployUrl": "http://<bucketName>.s3-website-eu-west-1.amazonaws.com/7edd45b6ad71be8feb0ff6a43a13a0f8/index.html",
}
`;
